### ISSUE TITLE
Tests with the Explicit attribute are run when all test case filters are using the operator !~

### ISSUE THREAD

**Issue by dmurty** [nunit/nunit3-vs-adapter#1078]
Nunit version 3.13.3
Nunit3TestAdpater version  4.3.1.0
Visual Studio 2022 17.5.0

We are using TeamCity Professional 2022.10.2 (build 117025) and using a dotnet runner to run tests.
They have a nice feature that will automatically split tests up to run in parallel on different build agents. It does this using testCaseFilters

The catch all case runs tests that were not run on the other agents. It does this by building a .runsettings file that contains a bunch of 
FullyQualifiedName!~ testcasefilters in the file. Unfortunately when run, the nunit3TestAdapter runs with Explicit run.

I attached a copy of the .runsettings file that teamcity generated.
[CatchAll.runsettings.txt](https://github.com/nunit/nunit3-vs-adapter/files/10976410/CatchAll.runsettings.txt)

Here is the log from Teamcity that shows that its running Explicit tests

dotnet test
  Starting: .NET SDK 7.0.201 "C:\Program Files\dotnet\dotnet.exe" test C:\BuildAgent\work\9dbf0d877fbc14df\NET\Bin\Release\HealthCareSynergyPatientsChartsTest.dll --settings C:\BuildAgent\temp\agentTmp\13.runsettings /logger:logger://teamcity /TestAdapterPath:C:\BuildAgent\plugins\dotnet\tools\vstest15 /logger:console;verbosity=normal -- NUnit.ExplicitMode=Strict
  in directory: C:\BuildAgent\work\9dbf0d877fbc14df
  Microsoft (R) Test Execution Command Line Tool Version 17.5.0 (x64)
  Copyright (c) Microsoft Corporation.  All rights reserved.
  
  Starting test execution, please wait...
  A total of 1 test files matched the specified pattern.
  NUnit Adapter 4.3.1.0: Test execution started
  Running selected tests in C:\BuildAgent\work\9dbf0d877fbc14df\NET\Bin\Release\HealthCareSynergyPatientsChartsTest.dll
     NUnit3TestExecutor discovered 92 of 92 NUnit test cases using Current Discovery mode, Explicit run
  HealthCareSynergyPatientsChartsTest: HealthCareSynergyPatientsChartsTest.Classes.ClientCleanupTesting.TestChangeCancelledDocumentToInProgress
    Failed TestChangeCancelledDocumentToInProgress [1 s]
    Error Message:
     HSIAccess.PassportException : Username or Password is invalid.

We would really like to use this Teamcity feature, but this issue is preventing us from doing so.

Please let me know how to fix this issue.

Thanks,

**Comment by dmurty** [comment]
It seems like if a test's Fully Qualified Name does not contain the filter text then Nunit is seeing as being explicitly run. Is that correct.

Teamcity ran all the tests on one build agent, and it did not run any tests that were marked as explicit. It then distributes the tests across multiple build agents using the testCaseFilter of FullyQualifiedName!~<list of test fixtures that was run on the base run>. But if this is now running explicit tests, this will never work. What should I do. We still want these test around to run if the a Developer explicitly runs that test.

**Comment by OsirisTerje** [comment]
When TeamCity ran **all tests**, those were not explicit.  When it lists something that can be regarded as a list of tests, which this 2nd round will be, then it actually **IS*  an explicit test run.  

There is nothing the adapter can do with this, and even if we had added a kind of "Never run explicit tests", the framework would still run them.  Adding such a feature to the framework is not likely to happen.  

There [at least] three ways you might go about this:
1. If you add another build step before the test where you add all the explicit tests to the list in the runsettings.  YOu can find these by running e.g. dotnet test --list-tests, this gives you all the tests, explicit and none, and if you collect the list of tests from the run all test, which excludes the explicit ones, subtract these and you get the list of explicit tests.  
2. Write your own little app that simple runs in the CI through all the tests and looks for the explicit attribute, and adds them to the list.
3. Write your own little cmd app to be run in the CI before compiling, which replaces all Explicit attributes with the Ignore attribute for the subsequent compiling.  Could be written in python something like : 
```python
import os

# Define the directory to search in and the file extension to match
search_directory = "/path/to/directory" #or use current directory
file_extension = ".cs"

# Define the word to replace and the replacement word
search_word = "[Explicit]"
replacement_word = "[Ignore(\"Was explicit\"]"


# Walk through the directory and its subdirectories
for root, dirs, files in os.walk(search_directory):
    for file in files:
        # Check if the file has the specified extension
        if file.endswith(file_extension):
            file_path = os.path.join(root, file)
            
            # Read the file contents
            with open(file_path, "r") as f:
                contents = f.read()
                
            # Replace the word and write the new contents back to the file
            new_contents = contents.replace(search_word, replacement_word)
            with open(file_path, "w") as f:
                f.write(new_contents)
```

**Comment by dmurty** [comment]
The first run, dotnet test is run on all the test assemblies without any TestCaseFilter. Then after that completes, TeamCity now has a list of TestFixtures that ran with how long it took to run and divides them up in different batches based on duration and runs them on separate build agents. It does this by using the TestCaseFilter. It has a catch all batch that has a TestCaseFilter that contains all !~ (does not contain) for all the TestFixtures it ran on the the other batches. It is this case that the TestAdapter is running the explicit tests. Basically it sounds like no one will be able to use this TeamCity feature if they use the Nunit3TestAdapter without writing their own work around.

Thanks,

**Comment by OsirisTerje** [comment]
Yes, it might actually be possible.  Just chekced the return from the framework, and it might work.  Could then make two new states, Explicitmode.None and  ExplicitMode.Only.   

Good idea!

**Comment by OsirisTerje** [comment]
I notice that it says ExplicitMode=Strict and you seems to run mixed explict and non-explicit tests, and that doesnt make sense.  I believe the re might be a bug here too, and I think it may be related to the negative list, that is a list of tests not to be run.  That might probably confuse the logic there somewhere,  Anyway, this is both a good idea, and possibly a bug too.

**Comment by dmurty** [comment]
Thank you for looking further into this. Appreciate it.

Thanks again.

**Comment by OsirisTerje** [comment]
I have made a repro for this, and in both cases of running with a filter, either on command line or in a runsettings file, it works as it should, that is, it does a Non-Explicit run, and the explicit tests are not being run.  

Link to the [repro of this](https://github.com/nunit/nunit3-vs-adapter.issues/tree/master/Issue1078)

Run the repro with 

```cmd
dotnet test -s .runsettings  --logger "console;verbosity=normal"
```
Expected result:  Test3 is run, but Test1, Test2 and TestEx (the explicit one) are not run.

![image](https://user-images.githubusercontent.com/203432/226102297-dbc3bd1f-673f-4a24-8181-09abfbd887c7.png)



alternatively, run

```cmd
dotnet test --filter FullyQualifiedName!~Test1  --logger "console;verbosity=normal"
```
Expected result:  Test2 and Test3 are run, and Test1 and TestEx are not run. 
![image](https://user-images.githubusercontent.com/203432/226102357-ac659116-1ac2-409e-8ad8-c846e61a283b.png)

So that indicates that the tests you have in your reported list are all explicit tests.  @dmurty Can you please check that?
If they are actually not that, could you please work on the repro I have linked and make that reproduce your case?

So there is no bug in the explicit check here, however, the addition of a mode for not running any explicit tests at all could be added, but it would be better if Team City was informed of this and filtered out test runs consisting of all explicit tests.  
Alternatively, you could modify the runsettings to always include one non-explicit test. 


PS:
1. The use of this setting TeamCity is using in the runsettings file is an undocumented feature, ref [1](https://github.com/MicrosoftDocs/visualstudio-docs/issues/3968), and [2](https://github.com/MicrosoftDocs/visualstudio-docs/pull/6139)
2. The settings will not affect Visual Studio Test Explorer, there is a [reported and ignored](https://developercommunity.visualstudio.com/t/test-explorer-window-exposes-and-runs-tests-that-s/1261971) bug raised on that.

**Comment by OsirisTerje** [comment]
@dmurty Have you looked more into this?

**Comment by dmurty** [comment]
I have not yet. Unfortunately, work has been really busy with an April 1 deadline. I will try to take a look at it sometime this week.

 

Thanks,

 

Damien

**Comment by OsirisTerje** [comment]
I'll close this for now.

**Comment by OsirisTerje** [comment]
I do believe Jetbrains are aware of the issue now. I think there is another issue here with a link to their (cant find at the moment).
They need to fix this

But a mode that completely excludes explicits can be added, that require a new config setting though.  You can raise a separate issue for that.

**Comment by dmurty** [comment]
Sure, I will do that.

thanks,
Damien
________________________________
